### PR TITLE
[WIP][SQL] Optimize fromJavaDate() and toJavaDate()

### PR DESCRIPTION
**Pull request by MaxGekk**
### Why are the changes needed?
Before:
```
Save dates to ORC:                        Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
after 1582, noop                                   8410           8410           0         11.9          84.1       1.0X
before 1582, noop                                  8408           8408           0         11.9          84.1       1.0X
after 1582                                        15507          15507           0          6.4         155.1       0.5X
before 1582                                       15099          15099           0          6.6         151.0       0.6X

Load dates from ORC:                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
after 1582, vec off                               26858          27479         727          3.7         268.6       1.0X
```

After:
```
Save dates to ORC:                        Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
after 1582, noop                                   7518           7518           0         13.3          75.2       1.0X
before 1582, noop                                  7292           7292           0         13.7          72.9       1.0X
after 1582                                        13286          13286           0          7.5         132.9       0.6X
before 1582                                       16213          16213           0          6.2         162.1       0.5X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_231-b11 on Mac OS X 10.15.3
Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz
Load dates from ORC:                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
after 1582, vec off                               17867          17918          65          5.6         178.7       1.0X
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon This optimization requires to restore ugly code from Spark 2.4. I am not sure that it is worth

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 1]
```diff
@@ -84,6 +84,13 @@ object DateTimeUtils {
     tf.format(us)
   }
 
+  private def millisToDays(millisUtc: Long, timeZone: TimeZone): SQLDate = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:87]: shall we use `microsToDays`?

File: sql/core/benchmarks/DateTimeRebaseBenchmark-jdk11-results.txt [hunk 2]
```diff
@@ -59,36 +59,36 @@ OpenJDK 64-Bit Server VM 11.0.6+10-post-Ubuntu-1ubuntu118.04.1 on Linux 4.15.0-1
...
 
 OpenJDK 64-Bit Server VM 11.0.6+10-post-Ubuntu-1ubuntu118.04.1 on Linux 4.15.0-1063-aws
 Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
 Load dates from ORC:                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
-after 1582, vec off                               20874          20905          38          4.8         208.7       1.0X
-after 1582, vec on                                 3813           3844          28         26.2          38.1       5.5X
-before 1582, vec off                              25912          25949          38          3.9         259.1       0.8X
-before 1582, vec on                                4322           4343          19         23.1          43.2       4.8X
+after 1582, vec off                               14060          14251         193          7.1         140.6       1.0X
```
Comment (MaxGekk) [inline sql/core/benchmarks/DateTimeRebaseBenchmark-jdk11-results.txt:71]: @cloud-fan Read results look good.

File: sql/core/benchmarks/DateTimeRebaseBenchmark-jdk11-results.txt [hunk 3]
```diff
@@ -59,36 +59,36 @@ OpenJDK 64-Bit Server VM 11.0.6+10-post-Ubuntu-1ubuntu118.04.1 on Linux 4.15.0-1
 Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
 Save dates to ORC:                        Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
-after 1582, noop                                   9295           9295           0         10.8          93.0       1.0X
-before 1582, noop                                  9352           9352           0         10.7          93.5       1.0X
-after 1582                                        17112          17112           0          5.8         171.1       0.5X
-before 1582                                       17979          17979           0          5.6         179.8       0.5X
+after 1582, noop                                   9250           9250           0         10.8          92.5       1.0X
+before 1582, noop                                  9522           9522           0         10.5          95.2       1.0X
+after 1582                                        16377          16377           0          6.1         163.8       0.6X
```
Comment (MaxGekk) [inline sql/core/benchmarks/DateTimeRebaseBenchmark-jdk11-results.txt:64]: @cloud-fan I don't think it worth to optimize the write path. The code looks ugly, from my point of view.

**Comment by MaxGekk** [comment]
I am closing this PR because the final results look not so good:
```diff
OpenJDK 64-Bit Server VM 1.8.0_242-8u242-b08-0ubuntu3~18.04-b08 on Linux 4.15.0-1063-aws
 Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
 Load dates from ORC:                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
-after 1582, vec off                               39651          39686          31          2.5         396.5       1.0X
-after 1582, vec on                                 3647           3660          13         27.4          36.5      10.9X
-before 1582, vec off                              38155          38219          61          2.6         381.6       1.0X
-before 1582, vec on                                4041           4046           6         24.7          40.4       9.8X
+after 1582, vec off                               76436          77047         877          1.3         764.4       1.0X
+after 1582, vec on                                 3790           3797          10         26.4          37.9      20.2X
+before 1582, vec off                              52369          52460         105          1.9         523.7       1.5X
+before 1582, vec on                                4171           4182          10         24.0          41.7      18.3X
```

**Pull request closed without merging**